### PR TITLE
Backend API Refactor

### PR DESCRIPTION
**Pull request by crockeea**
*Issue #, if available:*

*Description of changes:*
The primary goal of this PR is to create an internal API for the HE backend (SEAL) so that evaluators mostly call an inner API rather than SEAL directly. The purpose of this new API is to facilitate adding a different HE backend to HIT. It's a bit difficult to see the effects of this refactor in this PR alone, but these changes dramatically reduce the number of files touched by changing to the Lattigo backend. 

For example, now instead of using SEAL-specific functions inside the tests, tests use the `HEContext` API, and therefore don't need to be updated when the backend is changed. Similarly, most of the logging statements in `DebugEvaluator` that print HE parameters used to rely on SEAL-specific APIs; now they rely only on the generic `HEContext` APIs, and therefore don't need to be changed when moving to Lattigo.

Backend-specific functions are now isolated to
- `HEContext`
- `Ciphertext`
- `HomomorphicEvaluator`
- `DebugEvaluator` (minor)
- `ScaleEstimator` (minor)

Specific changes:
- Updated to latest SEAL release
- Move functions from `sealutils` into the new `HEContext` API, which is intended to capture backend-specific functionality.
- Remove unused imports
- Update a few miscellaneous comments, fix typos, and rename a few variable.
- Made a few `Evaluator` API functions _not_ `const` anymore. This is unfortunate, but necessary since the functions can't be `const` with Lattigo.
- Use SEAL `EncryptionParameter` serialization. Unclear why I wasn't using this before, but it simplifies instance serialization and makes the process more similar to how it works with Lattigo.
- Fix `enable-clang-format.cmake` script to properly report when clang-format isn't found.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/hit/api/context.cpp [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    vector<int> gen_modulus_vec(int num_primes, int log_scale) {
+        vector<int> modulusVector(num_primes);
+        // the SEAL examples recommend the last modulus be 60 bits; it's unclear why,
+        // and also unclear how closely that choice is related to log_scale (they use 40 in their examples)
+        modulusVector[0] = 60;
+        for (int i = 1; i < num_primes - 1; i++) {
+            modulusVector[i] = log_scale;
+        }
+        // The special modulus has to be as large as the largest prime in the chain.
+        modulusVector[num_primes - 1] = 60;
```
Comment (bryce-shang) [inline src/hit/api/context.cpp:28]: Nit:
* Input validation: may need check if num_primes > 0.
* More concise version
```
vector<int> modulusVector(num_primes, log_scale);
modulusVector[0] = 60;
modulusVector[num_primes - 1] = 60;
```

File: src/hit/api/context.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+        int max_ciphertext_level() const;
+        int num_slots() const;
+        uint64_t get_qi(int he_level) const;
+        uint64_t get_pi(int i) const;
+        int num_qi() const;
+        int num_pi() const;
+        uint64_t total_modulus_bits() const;
+        int min_log_scale() const;
+        int log_scale() const;
```
Comment (bryce-shang) [inline src/hit/api/context.h:30]: Nit: add some docs on these APIs?

File: tests/api/linearalgebra/linearalgebra.cpp [hunk 3]
```diff
@@ -1984,7 +1983,7 @@ TEST(LinearAlgebraTest, RescaleToNextMatrix) {
     EncryptedMatrix ct_mat2 = linear_algebra.multiply_plain(ct_mat1, 3.14);
 
     ASSERT_EQ(pow(2, LOG_SCALE * 2), ct_mat2.scale());
-    uint64_t prime = get_last_prime(ckks_instance.context, ct_mat1.he_level());
+    uint64_t prime = ckks_instance.context->get_qi(ct_mat1.he_level());
```
[the remark this answers is not in the document]
Comment (crockeea) [reply tests/api/linearalgebra/linearalgebra.cpp:1986]: Added comment in context.h where these functions are defined.

File: src/hit/api/linearalgebra/encryptedcolvector.h [hunk 4]
```diff
@@ -47,10 +47,10 @@ namespace hit {
         // use `encrypt_row_vector` in `LinearAlgebra` to construct an encrypted row vector
         EncryptedColVector() = default;
         // Returns a EncryptedColVector, which is deserialized from protobuf::EncryptedColVector.
-        EncryptedColVector(const std::shared_ptr<seal::SEALContext> &context,
+        EncryptedColVector(const std::shared_ptr<HEContext> &context,
                            const protobuf::EncryptedColVector &encrypted_col_vector);
         // Returns a EncryptedColVector, which is deserialized from a stream containing a protobuf::EncryptedColVector.
-        EncryptedColVector(const std::shared_ptr<seal::SEALContext> &context, std::istream &stream);
+        EncryptedColVector(const std::shared_ptr<HEContext> &context, std::istream &stream);
```
[the remark this answers is not in the document]
Comment (crockeea) [reply src/hit/api/linearalgebra/encryptedcolvector.h:53]: I think versioning would be good for HIT, but we don't have the resources for it currently.

File: src/hit/api/context.cpp (hunk 1 above)
Comment (crockeea) [reply src/hit/api/context.cpp:28]: Fixed

File: src/hit/api/context.h (hunk 2 above)
Comment (crockeea) [reply src/hit/api/context.h]: Done

**Review by bryce-shang — APPROVED** [review]

**Pull request merged**